src/core/server/base.js

```diff
@@ -39,10 +39,17 @@ import {
 
 import WebpackIsomorphicToolsConfig from './webpack-isomorphic-tools-config';
 
-export const createHistory = ({ req }) => {
-  return addQueryParamsToHistory({
-    history: createMemoryHistory({ initialEntries: [req.url] }),
-  });
+export const createHistory = ({ req, _log = log }) => {
+  let history;
+
+  try {
+    history = createMemoryHistory({ initialEntries: [req.url] });
+  } catch (err) {
```

Comment:
Catching and recovering from this error seems like it would leave the app in a very broken state -- the history will not have a starting URL. Really, the upstream library needs a fix. Until then I think we should let it throw and trigger a 500.

However, throwing this error does not trigger a 500 (it's an uncaught promise and thus a gateway timeout). This is the real bug; the patch here needs to fix that. All your current patch does is prevent one kind of error from getting thrown. If another one gets thrown, we will have an uncaught promise again.

We need a try/catch in [`app.use()`](https://github.com/mozilla/addons-frontend/blob/adb2ad0132798af9f5fee9e5c98c542265ba5b40/src/core/server/base.js#L243) so that it looks more like:

```js
  app.use(async (req, res, next) => {
    try {
      if (isDevelopment) {
        // ...
      }
      // Everthing else in the handler...
    } catch (handlerError) {
      log.error(`Caught an unexpected error while handling the request: ${handlerError}`);
      return next(handlerError);
    }
  });
```

After that change, I get an immediate 500 response, no uncaught promise, and I see a stack trace pointing to the error.

But, wait, why don't I see our generic 500 page? 

Here lies the second bug we need to fix. The error handler encountered an error while processing the error. Let's guard against that in [the error handler](https://github.com/mozilla/addons-frontend/blob/adb2ad0132798af9f5fee9e5c98c542265ba5b40/src/core/server/base.js#L387) like this:

```js
  app.use((error, req, res, next) => {
    try {
      // All error handler code...
      return showErrorPage({ createStore, error, status: 500, req, res, config });
    } catch (recoveryError) {
      log.error(`Additionally, the error handler caught an error: ${recoveryError}`);
      log.error({ err: recoveryError }); // log the stack trace.

      // Pass the original error to the next error handler.
      return next(error);
    }
  });
```

With this change, I see the first stack trace, the one we needed to know about all along.

Reply:
Thanks @kumar303, you're right it is better to fix the root cause than patching the visible problem. My mind was not fresh enough, I'll revisit this patch by tomorrow.

Reply:
> Catching and recovering from this error seems like it would leave the app in a very broken state -- the history will not have a starting URL. Really, the upstream library needs a fix. Until then I think we should let it throw and trigger a 500.

The history had `/`, which is enough to show a nice error page. Any navigation will trigger a `LOCATION_CHANGE`, so the state is not super broken, only the location is not correct but that's because the URL is not valid. That's the main issue and not sure the upstream lib can do anything about it.

Reply:
> But, wait, why don't I see our generic 500 page?
> 
> Here lies the second bug we need to fix. The error handler encountered an error while processing the error. Let's guard against that in [the error handler](https://github.com/mozilla/addons-frontend/blob/adb2ad0132798af9f5fee9e5c98c542265ba5b40/src/core/server/base.js#L387) like this:

I still don't see any change, the nice error page is not displayed because we need the store, which needs a history. I wish we had static error pages, at least for 5xx.

Reply:
>> Here lies the second bug we need to fix. The error handler encountered an error while processing the error.

> I still don't see any change

The stack traces are different (but they are hard to read because of the async generator implementation).

Without protecting the error handler, the stack trace points to the error handler as the culprit:

https://github.com/mozilla/addons-frontend/blob/eaf095d9631c21ac98533fc56f00d981373d0cc2/src/core/server/base.js#L382

...and then `createHistory()`:

https://github.com/mozilla/addons-frontend/blob/eaf095d9631c21ac98533fc56f00d981373d0cc2/src/core/server/base.js#L101

But the real culprit was in the request handler:

https://github.com/mozilla/addons-frontend/blob/eaf095d9631c21ac98533fc56f00d981373d0cc2/src/core/server/base.js#L267

The code example I provided above puts a focus on the real stack trace but logs the side effect stack trace, too.